Fix reauth flow, improve auth UI, and fix subscription task cleanup

steynovich opened the pull request:
## Summary

- **Fix reauth flow creating duplicate config entries** - The reauth flow was incorrectly calling `async_create_entry()` instead of `async_abort()`, causing duplicate entries and "Invalid flow specified" errors
- **Improve cookie authentication flow UI** - Added step-by-step instructions, input validation, and specific error messages to make the credential setup process easier
- **Fix subscription task not cancelled on entry unload** - The subscription task wasn't being cancelled when unloading the config entry, causing `KeyError` exceptions

## Test plan

- [x] Verified tests pass (6 passed, 2 skipped, 1 pre-existing failure)
- [ ] Test reauth flow no longer creates duplicate entries
- [ ] Test improved config flow UI shows step-by-step instructions
- [ ] Test unloading integration no longer causes KeyError exceptions

Fixes #466

steynovich committed:
Fix reauth flow creating duplicate config entries

The reauth flow was incorrectly calling async_create_entry() after
updating the existing config entry, which caused Home Assistant to
show "Invalid flow specified" errors and create duplicate entries.

Changed to properly call async_abort(reason="reauth_successful")
after updating credentials, which is the expected behavior for
reauth flows.

Also fixed the config entry migration code to use async_update_entry()
instead of direct assignment, which is required by newer Home Assistant
versions.

Fixes #466

steynovich committed:
Improve cookie authentication flow UI

- Add step-by-step instructions in the config flow description
- Add input validation before making API calls
- Add specific error messages for invalid issue_token and cookies format
- Add data_description hints showing expected input format
- Add reauth_successful translation to en.json

steynovich committed:
Fix subscription task not cancelled on entry unload

- Add subscription_task field to HomeAssistantNestProtectData
- Cancel task in async_unload_entry before removing data
- Add check in _async_subscribe_for_data to exit if entry unloaded
- Update _register_subscribe_task to track current task reference

This fixes the KeyError when accessing hass.data[DOMAIN][entry.entry_id]
after the config entry has been unloaded.

steynovich committed:
Fix test_get_first_data_success test

- Add missing weather_for_structures and 2fa_enabled fields to mock response
- Fix client.py to rename 2fa_enabled even when False (check key existence, not truthiness)
- Update test assertions to check FirstDataAPIResponse object properties

iMicknl commented:
Thanks for your contribution, @steynovich! Much appreciated. I will have a look asap.

Regarding the failing test, good to check if main has the same issues, as the project didn't get much love lately unfortunately.

steynovich committed:
Fix test suite issues

- Fix test_get_first_data_success: add missing API response fields and
  update assertions for FirstDataAPIResponse object
- Fix client.py: rename 2fa_enabled field even when False
- Add pytest.ini asyncio_default_fixture_loop_scope setting
- Remove deprecated event_loop fixture from tests
- Add pynest conftest.py to override auto fixtures

steynovich committed:
Override verify_cleanup fixture for pynest tests

iMicknl commented on custom_components/nest_protect/__init__.py:
```diff
@@ -116,39 +117,63 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Unload a config entry."""
+    # Cancel subscription task before unloading
+    if entry.entry_id in hass.data.get(DOMAIN, {}):
```
Is this additional check required? We only remove/pop the entry id from hass.data on line 150 right?

steynovich replied:
Yes, the check is still needed as a defensive measure for edge cases:

1. **Failed setup** - If `async_setup_entry` failed before adding the entry to `hass.data[DOMAIN]`, the unload could still be called
2. **Race conditions** - Another unload attempt could have already removed it
3. **Prevents KeyError** - Without it, accessing `hass.data[DOMAIN][entry.entry_id]` would raise `KeyError`

It's a common Home Assistant pattern to guard against these scenarios.

steynovich committed:
Address PR review feedback from Copilot

- Improve issue token URL validation (check for query params)
- Enhance cookie validation (check for key-value pairs and Google auth markers)
- Fix task cleanup order: cancel subscription after successful platform unload
- Store stripped credentials back into user_input before API calls
- Add CancelledError handling to prevent orphaned subscription tasks
- Add explanatory comments for exception handling

iMicknl reviewed (approved):
Thanks!!
